packages/backend/src/controllers/authentication.ts

```diff
@@ -499,28 +531,193 @@ export const isOktaPassportStrategyAvailableToUse = !!(
+        azuread.openIdConnectMetadataEndpoint!,
+    );
+
+    /**
+     * Load the JWK from a file or directly from configuration:
+     */
+    const jwk =
+        (azuread.certificateKeyJwkJson &&
+            loadJWKFromConfig(azuread.certificateKeyJwkJson)) ||
+        (azuread.certificateKeyJwkPath &&
+            (await loadJWKFromPath(azuread.certificateKeyJwkPath)));
+
+    if (!jwk) {
+        throw new Error(
+            'invariant: Could not find a valid JWK when setting up private_key_jwt',
+        );
+    }
+
+    /**
+     * Azure expects the key's identifier (kid) to be the same value as the
+     * certificate SHA-1 thumbprint, base64-encoded, for purposes of key
+     * matching.
+     *
+     * Confusingly, it may also complain about the x5t and x5c claims, but we
+     * can ignore it here and simply override whatever value we have from the
+     * underlying private key, and node-openid-client will know to include it
+     * as part of the jwt header.
+     */
+    const certificateIdentifier = Buffer.from(
+        azuread.certificateThumbprint!,
+        'hex',
+    ).toString('base64url');
+
+    /**
+     * Build the JWKS out of the single private key:
+     */
+    const jwks = {
+        keys: [
+            {
+                ...jwk,
+                kid: certificateIdentifier,
+            },
+        ],
+    };
+
+    const client = new issuer.Client(
+        {
+            client_id: azuread.oauth2ClientId!,
+            token_endpoint_auth_signing_alg: 'RS256',
+            token_endpoint_auth_method: 'private_key_jwt',
+        },
+        jwks,
+    );
+
+    return new OpenIdClientStrategy(
+        {
+            client,
+            usePKCE: true,
+            passReqToCallback: true,
+            params: {
+                redirect_uri: new URL(
+                    `/api/v1${azuread.callbackPath}`,
+                    lightdashConfig.siteUrl,
+                ).href,
+            },
+            extras: {
+                clientAssertionPayload: {
+                    /**
+                     * AzureAD only expects a single audience value, but node-openid-client may
+                     * set more than one value here - we override this part of the payload to
+                     * make Azure happy:
+                     */
+                    aud: issuer.metadata.issuer,
+
+                    /** AzureAD complains if this is missing: */
+                    typ: 'JWT',
+                },
+            },
+        },
+
+        /**
+         * This is compatible, but types differ from what's otherwise expected.
+         */
+        genericOidcHandler(
+            OpenIdIdentityIssuerType.AZUREAD,
+            new URL(azuread.openIdConnectMetadataEndpoint!).origin,
+        ) as unknown as StrategyVerifyCallback<unknown>,
+    );
+};
+
+/**
+ * Creates the appropriate AzureAd passport strategy based on the available configuration.
+ */
+export const createAzureAdPassportStrategy = () => {
```

Comment:
To avoid unintended issues, I kept both strategies around, and only use the new strategy if using `private_key_jwt`. In theory they should behave the same, and we can migrate other issuers to also use `openid-client`.